src/prefect/engine/flow_runner.py

```diff
@@ -432,17 +456,88 @@ def get_flow_run_state(
                         result=ConstantResult(value=val),
                     )
 
-                # -- run the task
+                # handle mapped tasks
+                if any([edge.mapped for edge in upstream_states.keys()]):
 
-                with prefect.context(task_full_name=task.name, task_tags=task.tags):
+                    ## wait on upstream states to determine the width of the pipeline
+                    ## this is the key to depth-first execution
+                    upstream_states.update(
+                        executor.wait(
+                            {e: state for e, state in upstream_states.items()}
+                        )
+                    )
+
+                    ## we submit the task to the task runner to determine if
+                    ## we can proceed with mapping - if the new task state is not a Mapped
+                    ## state then we don't proceed
+                    task_states[task] = executor.wait(
+                        executor.submit(
+                            self.run_task,
+                            task=task,
+                            state=task_state,  # original state
+                            upstream_states=upstream_states,
+                            context=dict(
+                                prefect.context, **task_contexts.get(task, {})
+                            ),
+                            task_runner_state_handlers=task_runner_state_handlers,
+                            upstream_mapped_states=upstream_mapped_states,
+                            mapped_parent=True,
+                        )
+                    )
+
+                    ## either way, we should now have enough resolved states to restructure
+                    ## the upstream states into a list of upstream state dictionaries to iterate over
+                    list_of_upstream_states = prepare_upstream_states_for_mapping(
+                        task_states[task], upstream_states, mapped_children
+                    )
+
+                    submitted_states = []
+
+                    for idx, states in enumerate(list_of_upstream_states):
+                        ## if we are on a future rerun of a partially complete flow run,
```

Comment:
Following our debugging session, this is definitely the correct comment. However I think in a few months we won't understand how/why it translates into the if statement below it